components/org.wso2.micro.integrator.initializer/src/main/java/org/wso2/micro/integrator/initializer/handler/transaction/TransactionCountHandler.java

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+*
+* WSO2 Inc. licenses this file to you under the Apache License,
+* Version 2.0 (the "License"); you may not use this file except
+* in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied. See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package org.wso2.micro.integrator.initializer.handler.transaction;
+
+import org.apache.axis2.transport.base.BaseConstants;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.AbstractSynapseHandler;
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.core.axis2.Axis2MessageContext;
+import org.wso2.micro.integrator.initializer.handler.DataHolder;
+import org.wso2.micro.integrator.initializer.handler.transaction.store.TransactionStore;
+
+public class TransactionCountHandler extends AbstractSynapseHandler {
+
+    private static final Log LOG = LogFactory.getLog(TransactionCountHandler.class);
+    private static int transactionCount = 0;
+
+    @Override
+    public boolean handleRequestInFlow(MessageContext messageContext) {
+        org.apache.axis2.context.MessageContext axis2MessageContext =
+                ((Axis2MessageContext) messageContext).getAxis2MessageContext();
+        Object transactionProperty = axis2MessageContext.getProperty(BaseConstants.INTERNAL_TRANSACTION_COUNTED);
+        // increment the transaction count by 1 if the INTERNAL_TRANSACTION_COUNTED message property has not been set
+        // or if the value of the property is set to false.
+        if (!(transactionProperty instanceof Boolean && (Boolean) transactionProperty)) {
+            axis2MessageContext.setProperty(BaseConstants.INTERNAL_TRANSACTION_COUNTED, true);
+            // commit the current transaction count to database and set the transaction count to zero if it has
+            // reached to Interger.MAX_VALUE
+            commitAndResetTransactionCountIfReachedToMaxIntegerValue();
```

Comment:
Shall we execute this periodically , with some lesser period. True that it won't be 100% accurate but executing this ( checking int max value ) for each and every request might affect performance in other hand . @sdkottegoda

Reply:
Checking against int max will not cause a considerable degradation. But, we can avoid it by making the transaction handling part asynchronous.

Reply:
However, I just realized that this method is one that will definitely be called concurrently and we need to have some synchronization in updating the counter.

So I think, we need to notify an async Counter thread which will update the transaction count in a synchronized method. @sajinieKavindya

Reply:
I think the performance effect this simple check can cause if negligible when compared to all the other processings done through the message flow of a request. And if we run this periodically, there is a chance that the transaction count gets exceeded the int max value by that time this value is checked, which will cause some issues afterward. But if we are going to check the count with a grace value as suggested in https://github.com/wso2/micro-integrator/pull/1537#discussion_r425904061, maybe we can eliminate that issue. But it again depends on how often we are going to execute this check. (Therefore, this update interval and the grace value needs to be chosen wisely in that case). @sdkottegoda

Reply:
True it is of negligible performance impact. But IMO better to avoid them wherever possible. However moving to an async thread will sort all these issues I think. Also shall we verify this with concurrent requests to ensure the level of correctness. ( To test concurrent requests you can simply do in a  single command with Apache benchmark )

Reply:
This issue will be addressed separately here https://github.com/wso2/micro-integrator/issues/1558